### PR TITLE
[BIOMAGE-877] Fixes for local dev

### PR DESCRIPTION
**Pull request by kafkasl**
Changes
* Download both processed-matrix & biomage-sources from S3 in pull command
* Downloading `projects` table from dynamoDB during pull command
* Using env variable to patch permissions for experiments during pull command.

More info on https://github.com/biomage-ltd/ui/pull/367

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -23,7 +23,7 @@ install: clean ## Creates venv, and adds biomage as system command
 
 	@echo "==> Installing utility and dependencies..."
 	@venv/bin/pip install --upgrade pip
-	@venv/bin/pip install .
+	@venv/bin/pip install -e .
```
Comment (ivababukova) [inline Makefile:26]: What does the `-e` flag do?

File: biomage/experiment/pull.py [hunk 2]
```diff
@@ -36,6 +41,10 @@ def download_if_modified(bucket, key, filepath):
     s3 = boto3.resource("s3")
     s3_obj = s3.Object(bucket, key)
 
+    try:
+        s3_obj.last_modified
```
Comment (ivababukova) [inline biomage/experiment/pull.py:45]: The try statement seems odd -- is the idea to check if `s3_obj.last_modified` will result in a clienterror?

File: biomage/experiment/pull.py [hunk 3]
```diff
@@ -129,13 +191,17 @@ def pull(experiment_id, input_env):
...
     bucket = f"biomage-source-{input_env}"
-    file = f"{experiment_id}/{RDS_FILE}"
-    dst_file = file + ".gz"
+    file = f"{experiment_id}/r.rds"
+    dst_file = f"{experiment_id}/{SOURCE_RDS_FILE}.gz"
+    download_if_modified(bucket=bucket, key=file, filepath=dst_file)
+
+    bucket = f"processed-matrix-{input_env}"
+    file = f"{experiment_id}/r.rds"
+    dst_file = f"{experiment_id}/{PROCESSED_RDS_FILE}.gz"
```
Comment (ivababukova) [inline biomage/experiment/pull.py:200]: what does `dst` stand for?

File: biomage/experiment/pull.py [hunk 4]
```diff
@@ -72,26 +106,54 @@ def update_config_if_needed(filepath, table_name, experiment_id):
...
+def update_configs(experiment_id, origin):
+    # update samples
+    file_path = f"{experiment_id}/{SAMPLES_FILE}"
+    table_name = f"{SAMPLES_TABLE}-{origin}"
+    project_uuid = update_config_if_needed(file_path, table_name, experiment_id)
+
+    # update experiments
+    file_path = f"{experiment_id}/{EXPERIMENTS_FILE}"
+    table_name = f"{EXPERIMENTS_TABLE}-{origin}"
+    project_uuid = update_experiment_config_if_needed(
```
Comment (ivababukova) [inline biomage/experiment/pull.py:149]: Why are you getting `project_uuid` assigned to both on line 144 and on line 149? Makes me worried what would happen if in the future we swap those two lines -- will that make `project_uuid` have a wrong value?

File: biomage/experiment/utils.py [hunk 5]
```diff
@@ -116,3 +118,35 @@ def download_S3_json(s3_obj, key, filepath):
     s3_obj.download_file(local_file)
 
     set_modified_date(file_location=local_file, date=s3_obj.last_modified)
+
+
+def get_cognito_username(email):
+    client = boto3.client("cognito-idp")
```
Comment (ivababukova) [inline biomage/experiment/utils.py:124]: [out of scope for this PR] Do you know if we can use this API and get the login times of a user? @aerlaut might be interested in looking into this for the user KPIs epic

File: biomage/experiment/utils.py [hunk 6]
```diff
@@ -116,3 +118,35 @@ def download_S3_json(s3_obj, key, filepath):
...
+    user_name = client.admin_get_user(Username=email, UserPoolId="eu-west-1_sVQL15Yxu")[
```
Comment (ivababukova) [inline biomage/experiment/utils.py:125]: Looks to me that the userPoolId should be added to the config yaml instead: https://github.com/biomage-ltd/biomage-utils/blob/master/config.yaml

File: biomage/experiment/utils.py [hunk 7]
```diff
@@ -116,3 +118,35 @@ def download_S3_json(s3_obj, key, filepath):
...
+        "Username"
+    ]
+
+    return user_name
+
+
+def add_user_to_rbac(user_name, cfg):
+    if "rbac_can_write" in cfg:
+        cfg["rbac_can_write"]["SS"].append(user_name)
```
Comment (ivababukova) [inline biomage/experiment/utils.py:134]: what would happen if we try to add a username that's already been added? Will it add it twice, or will the fact that the values in `rbac_can_write` are a set prevent from this to happen?

File: biomage/experiment/utils.py (hunk 6 above)
Comment (ivababukova) [reply biomage/experiment/utils.py:125]: Or maybe in constants.py?

File: Makefile (hunk 1 above)
Comment (robioson) [reply Makefile:26]: Installs with symlinks; it means you can develop utils without reinstalling after my recent packaging changes.

Comment (kafkasl) [reply Makefile:26]: it links the installed packages to the source code so you don't have you `make install` when you develop in order to make the changes available for `biomage` command.

File: biomage/experiment/utils.py [hunk 8]
```diff
@@ -116,3 +120,36 @@ def download_S3_json(s3_obj, key, filepath):
...
+    return user_name
+
+
+def add_user_to_rbac(user_name, cfg):
+    if "rbac_can_write" in cfg:
+        if user_name not in cfg["rbac_can_write"]["SS"]:
+            cfg["rbac_can_write"]["SS"].append(user_name)
+        return cfg
+    for val in cfg.values():
+        if isinstance(val, dict):
```
Comment (robioson) [inline biomage/experiment/utils.py:140]: This makes me suspicious, what else is in `cfg`?

**Pull request merged**